performance-tracking-core/src/main/java/com/rakuten/tech/mobile/perf/core/MeasurementBuffer.java

```diff
@@ -7,51 +7,102 @@
   // Size must by a power of 2 so that Integer.MAX_VALUE is divisible by it
   static final int SIZE = 512;
 
-  final Measurement[] at = new Measurement[SIZE];
-  final AtomicInteger nextTrackingId = new AtomicInteger(1);
+  private final Measurement[] measurements = new Measurement[SIZE];
+  // Tracking id shouldn't be zero as zero has special purpose
+  private final AtomicInteger nextTrackingId = new AtomicInteger(1);
 
   MeasurementBuffer() {
     for (int i = 0; i < SIZE; i++) {
-      at[i] = new Measurement();
+      measurements[i] = new Measurement();
     }
   }
 
+  // Constructor for testing only
+  MeasurementBuffer(int firstTrackingId) {
+    this();
+    nextTrackingId.set(firstTrackingId);
+  }
+
+  /**
+   * Retrieves measurement at the provided {@code index}
+   *
+   * @param index to retrieve
+   * @return the measurement for the provided index
+   */
+  Measurement at(int index) {
+    return measurements[index];
+  }
+
+  /**
+   * Adds the next tracking id to the next buffer slot and returns the measurement for that slot
+   *
+   * @return the next measurement or null if the buffer is full
+   */
   Measurement next() {
     int id = nextTrackingId.getAndIncrement();
+    int nextIndex = indexForId(id);
 
-    if (id == 0) {
-      // Ensure id is not zero as zero has special purpose
-      id = nextTrackingId.getAndIncrement();
+    if (isFull(id)) {
+      nextTrackingId.getAndDecrement();
+      return null;
     }
 
-    int index = id % SIZE;
-    if (index < 0) {
-      index += SIZE;
+    Measurement measurement = measurements[nextIndex];
+    measurement.trackingId = id;
+
+    return measurement;
+  }
+
+  /**
+   * Retrieves a measurement by tracking Id
+   *
+   * @param trackingId to retrieve
+   * @return measurement with the provided tracking id, or null if the tracking id doesn't exist
+   */
+  Measurement getByTrackingId(int trackingId) {
+    Measurement measurement = measurements[indexForId(trackingId)];
+    if (measurement.trackingId == trackingId) {
+      return measurement;
     }
 
-    Measurement m = at[index];
+    return null;
+  }
 
-    if (m.trackingId != 0) {
-      nextTrackingId.getAndDecrement();
-      return null;
+  /**
+   * Returns the count of measurements in the buffer from {@code startIndex}
+   *
+   * @param startIndex to count from
+   * @return count of measurements
+   */
+  int count(int startIndex) {
+    int id = nextTrackingId.get();
+    int endIndex = indexForId(id);
+
+    int count;
+
+    if (startIndex == endIndex && isFull(id)) {
+      count = SIZE;
+    } else {
+      count = endIndex - startIndex;
+      if (count < 0) {
+        count += SIZE;
+      }
     }
 
-    m.trackingId = id;
+    return count;
+  }
 
-    return m;
+  private boolean isFull(int nextId) {
```

Comment:
`isFull` is better than before but this method still name throws me off a bit - on call side it expresses the intent: we want to see if the buffer is full.
but on implementation side it does not check if the buffer is full at all - it only checks if the id is set in the buffer.

I also noticed that in both places where this is called we already have the `endIndex = indexForId(nextId)` just before the call so doing it in the method seems superfluous.

maybe instead of the positive check we could negate it and do a 
```
if (! isEmpty(index))
// etc.

private boolean isEmpty(int nextId) {
  return at(indexForId(nextId)).trackingId == 0;
}
```

that doesn't express the call side intent as well though... but it doesn't lie about the implementation...

or it could be a `isFullUntilIndex(index)`? that's a bit more verbose but expresses both call and implementation side intent relatively well.

What do you think?

- [ ] Re-evaluate method name one last time

Reply:
Hmm, maybe this method is needed at all. Some of the other changes made this class much easier to read, so I think just inlining `at(nextIndex).trackingId != 0` is okay.